projects/RabbitMQ.Client/client/impl/Frame.cs

```diff
@@ -42,127 +42,117 @@
+             * | 2 bytes  | 2 bytes  | 8 bytes           | x bytes   |
+             * +----------+----------+-------------------+-----------+ */
+            public const int FrameSize = BaseFrameSize + 2 + 2 + 8;
+
+            public static int WriteTo(Span<byte> span, ushort channel, ContentHeaderBase header, int bodyLength)
+            {
+                const int StartClassId = StartPayload;
+                const int StartWeight = StartPayload + 2;
+                const int StartBodyLength = StartPayload + 4;
+                const int StartHeaderArguments = StartPayload + 12;
+
+                NetworkOrderSerializer.WriteUInt16(span.Slice(StartClassId), header.ProtocolClassId);
+                NetworkOrderSerializer.WriteUInt16(span.Slice(StartWeight), 0); // Weight - not used
+                NetworkOrderSerializer.WriteUInt64(span.Slice(StartBodyLength), (ulong)bodyLength);
+                var headerWriter = new ContentHeaderPropertyWriter(span.Slice(StartHeaderArguments));
+                header.WritePropertiesTo(ref headerWriter);
+                return WriteBaseFrame(span, FrameType.FrameHeader, channel, StartHeaderArguments - StartPayload + headerWriter.Offset);
+            }
         }
 
-        internal override int GetMinimumPayloadBufferSize()
+        internal static class BodySegment
         {
-            return 0;
-        }
+            /* +--------------+
+             * | Body segment |
+             * +--------------+
+             * | x bytes      |
+             * +--------------+ */
+            public const int FrameSize = BaseFrameSize;
+
+            public static int WriteTo(Span<byte> span, ushort channel, ReadOnlySpan<byte> body)
+            {
+                const int StartBodyArgument = StartPayload;
 
-        internal override int WritePayload(Memory<byte> memory)
-        {
-            return 0;
+                body.CopyTo(span.Slice(StartBodyArgument));
+                return WriteBaseFrame(span, FrameType.FrameBody, channel, StartBodyArgument - StartPayload + body.Length);
+            }
         }
-    }
 
-    abstract class OutboundFrame : Frame
-    {
-        public int ByteCount { get; private set; } = 0;
-        public OutboundFrame(FrameType type, int channel) : base(type, channel)
+        internal static class Heartbeat
         {
-        }
+            /* Empty frame */
+            public const int FrameSize = BaseFrameSize;
 
-        internal void WriteTo(Memory<byte> memory)
-        {
-            memory.Span[0] = (byte)Type;
-            NetworkOrderSerializer.WriteUInt16(memory.Slice(1).Span, (ushort)Channel);
-            int bytesWritten = WritePayload(memory.Slice(7));
-            NetworkOrderSerializer.WriteUInt32(memory.Slice(3).Span, (uint)bytesWritten);
-            memory.Span[bytesWritten + 7] = Constants.FrameEnd;
-            ByteCount = bytesWritten + 8;
-        }
+            public static Span<byte> Payload => new byte[]
```

Comment:
Any reason we are using an expression bodied member here and not the same fixed byte array all the time? It seems we only copy it into the rented memory but don't modify it, unless I missed that part

Reply:
Reason is a compiler trick, see here: https://github.com/dotnet/roslyn/pull/24621
Yeah I just saw it got to be a ReadOnlySpan (which it should've been anyway) in order to make it fully work.

Reply:
That's one of the things I despise a bit about all those tricks because most mere mortals can't remember that stuff unless you use that type of magic every day.

Reply:
I'm slightly modify it and add some comments explaining it so others do understand it faster why it is this way.